src/language/mongoDBService.ts

```diff
@@ -544,146 +446,183 @@ export default class MongoDBService {
       this._cachedFields[namespace] = fields ? fields : [];
     }
   }
 
-  _clearCurrentSessionDatabases(): void {
-    this._cachedDatabases = [];
+  /**
+   * Convert database names to Completion Items and cache them.
+   */
+  _cacheDatabaseCompletionItems(databases: Document[]): void {
+    this._cachedDatabases = databases.map((item) => ({
+      label: (item as { name: string }).name,
+      kind: CompletionItemKind.Field,
+      preselect: true,
+    }));
   }
 
-  _updateCurrentSessionDatabases(databases: CompletionItem[]): void {
-    this._cachedDatabases = databases ? databases : [];
+  /**
+   * Convert collection names to Completion Items and cache them.
+   */
+  _cacheCollectionCompletionItems(
+    textFromEditor: string,
+    position: { line: number; character: number },
+    database: string,
+    collections: Document[]
+  ): void {
+    this._cachedCollections[database] = collections.map((item) => {
+      if (this._isValidPropertyName(item.name)) {
+        return {
+          label: item.name,
+          kind: CompletionItemKind.Folder,
+          preselect: true,
+        };
+      }
+
+      // Convert invalid property names to array-like format.
+      const filterText = textFromEditor.split('\n')[position.line];
+
+      return {
+        label: item.name,
+        kind: CompletionItemKind.Folder,
+        // Find the line with invalid property name.
+        filterText: [
+          filterText.slice(0, position.character),
+          `.${item.name}`,
+          filterText.slice(position.character, filterText.length),
+        ].join(''),
+        textEdit: {
+          range: {
+            start: { line: position.line, character: 0 },
+            end: {
+              line: position.line,
+              character: filterText.length,
+            },
+          },
+          // Replace with array-like format.
+          newText: [
+            filterText.slice(0, position.character - 1),
+            `['${item.name}']`,
+            filterText.slice(position.character, filterText.length),
+          ].join(''),
+          preselect: true,
+        },
+      };
+    });
   }
 
-  _clearCurrentSessionCollections(): void {
-    this._cachedCollections = {};
+  _clearCachedFields(): void {
+    this._cachedFields = {};
   }
 
-  _updateCurrentSessionCollections(
-    database: string,
-    collections: CollectionItem[]
-  ): void {
-    if (database) {
-      this._cachedCollections[database] = collections ? collections : [];
-    }
+  _clearCachedDatabases(): void {
+    this._cachedDatabases = [];
+  }
+
+  _clearCachedCollections(): void {
+    this._cachedCollections = {};
   }
 
-  _clearCurrentSessionConnection(): void {
+  async _clearCurrentConnection(): Promise<void> {
+    this._connectionId = undefined;
     this._connectionString = undefined;
     this._connectionOptions = undefined;
+
+    if (this._serviceProvider) {
+      await this._serviceProvider.close(true);
+      this._serviceProvider = undefined;
```

Comment:
Is there any chance that something could call `connectToServiceProvider()` while the `close()` call is active? If so, you’ll probably need to do this to avoid a race condition where `connectToServiceProvider()` finishes before `close()` and you’d accidentally be overriding a _new_ value of `this._serviceProvider` here:

```suggestion
      const serviceProvider = this._serviceProvider;
      this._serviceProvider = undefined;
      await serviceProvider.close(true);
```